### ISSUE TITLE
[REQUEST] please cache the kakscript parser

### ISSUE THREAD

**Issue by mralusw** [mawww/kakoune#4435]
### Feature

It seems the parser caches none of the `def`, `eval` or `try` kakscript codeblocks. The parser merrily re-parses everything on evaluation.

You might say "don't stress / rely on kakscript so heavily", but this misses something. All builtin and external `kak` defs are evaluated all the time we interact with `kak`. The entire editor is slower (and consumes more battery) because the parser always parses. Over time, the difference accumulates at the exact ratio demonstrated below. The following code just highlights the inefficiency.

In other words, `kak` is just as efficient as the code you may not endorse. Just a few comments trigger a `40%` slowdown.

```
# timing commands from https://gitlab.com/kstr0k/k9s0ke-dbg.kak.git
def localtime-ns-save -params 0..2 -docstring %{
  localtime-ns-save [int-opt-name] [ -add | -remove | -negate ] [ SCOPE ]
  Save current timestamp (UNIX epoch) to %arg{1}, or k9s0ke_dbg_tstamp
} %{
  eval %sh{
    set -u
    var=${1:-k9s0ke_dbg_tstamp}; op=${2:-}; scope=${3:-global}
    sign=; if [ -negate = "$op" ]; then op=; sign=-; fi
    t=${EPOCHREALTIME:-}; t=${t%%.*}${t#*.}
    : "${t:=$(date +%s%N)}"
    printf 'set %s %s %s %s' "$op" "$scope" "$var" "$sign$t"
  }
} -override
def time-command -params .. -docstring %{
  Time the execution of command %arg{@}
  Result: %opt{k9s0ke_dbg_tstamp_diff}
  E.g. time-command eval %{nop %sh{sleep 1}}
  %arg{@} must be a command with args, not a string to be eval'd
} %{
  localtime-ns-save k9s0ke_dbg_tstamp_diff -negate
  %arg{@}
  localtime-ns-save k9s0ke_dbg_tstamp_diff -add
} -override
def time-command-dbg -params .. -docstring %{
  time-command %arg{@}, then print duration to *debug*
} %{
  time-command %arg{@}
  echo -debug -quoting kakoune "Timer: %opt{k9s0ke_dbg_tstamp_diff}" ( %arg{@} )
} -override


# looping support
def loop-32 -params .. %{
%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@}
%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@}
%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@}
%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@};%arg{@}
} -override -hidden


# parser test
def waste1 nop -override
def waste2 %{
  nop
  # comment
  # comment
  # comment
  # comment
  # comment
  # comment
  # comment
  # comment
} -override

# let's run .. twice
time-command-dbg loop-32 loop-32 loop-32 loop-32 waste1
time-command-dbg loop-32 loop-32 loop-32 loop-32 waste2
time-command-dbg loop-32 loop-32 loop-32 loop-32 waste1
time-command-dbg loop-32 loop-32 loop-32 loop-32 waste2

# press ccc in the *debug* buffer to visually separate runs
```

### Usecase

interpreter performance

**Comment by lenormf** [comment]
The editor uses so much RAM as it is, I can’t imagine what it would be like if all of `rc` was kept in memory. Maybe code from the modules loaded with `require-module` could be cached?

What about scripts sourced arbitrarily and commands run from the prompt (that use `eval`/`try`), would they also remain in RAM ad infinitum?

**Comment by eraserhd** [comment]
@lenormf - usually parsed scripts take less memory.  On the other hand, the language is not really designed to be fully tokenized.  As a shell-like language, there are all the problems with pre-parsing and tokenizing shell-like languages, including stuff like:

```
if [[ "$1" = x ]]; then
  foo=fi
  bar=:
else
  foo=:
  bar=fi
fi
if [[ x = y ]]; then
  printf 'one\n'
$foo
  printf 'two\n'
$bar
```

There's two valid parses here, and it can change at run time.  This same kind of issue could happen with, say, Kakoune's `try` statement.

**Comment by mralusw** [comment]
Codeblocks that are constantly being used should absolutely have their parsetrees in RAM.

It's quite simple to implement a [LRU cache](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)). This evicts cached artifacts once they are no longer "recently" used, and limits total memory used for the cache. There is a wealth of information and algorithms on caching, but even the simplest would do.

The cache does not need to care where the code comes from (`def`, `eval`, `try`, whether from `kakrc` or from loaded plugins). It can be configured conservatively (few cache entries), liberally (use up more memory), or user-defined.

But even a conservative setup would greatly speed any op.

**Comment by mralusw** [comment]
> On the other hand, the language is not really designed to be fully tokenized. As a shell-like language, there are all the problems with pre-parsing and tokenizing 

I've thought about that; but with kakscript, would there be any code fragments that can potentially lead to a different pasetree? I mean a shallow parse, not one that goes into `eval` blocks and guesses.

**Comment by Calum-J-I** [comment]
I'd be surprised to see this change happens. Seems like a lot of complexity to fix a use case that probably goes against mawww's idea of what kakscript should be. I suspect if he intended that people write enough kakscript for it to have a noticeable performance impact the language wouldn't be so deliberately limited. 

I also think you need to do some profiling where the command actually does some work. In this example you're running a _lot_ of commands that do very little work. In practice I think we're much more likely to be running a small number of commands where most of the time is spent running regex, manipulating selections, changing buffer contents, expanding `%sh` blocks or waiting on IO.

**Comment by Screwtapello** [comment]
What is a "parse tree" in this instance? It can't be a LISP-like AST, because Kakoune relies so heavily on expansions that any non-trivial AST would be invalidated near-instantly.

I can imagine that a command that stores a fixed string might possibly also cache token spans within that string... but that really only works for `define-command` that creates a long-lived command object, `try` and `eval` have no such persistent place to store a token cache.

You could solve *that* problem by making a hashmap mapping arbitrary strings to a cached tokenisation of those strings, but given that most such strings are mere dozens of bytes, hashing them and looking them up might well be slower than just re-tokenising them. Even if it worked, you'd then have spooky-action-at-a-distance performance problems where some operation was beautifully fast until it got evicted from the cache and then became horribly slow.

**Comment by mralusw** [comment]
> What is a "parse tree" in this instance? It can't be a LISP-like AST, because Kakoune relies so heavily on expansions that any non-trivial AST would be invalidated near-instantly.

Why? Consider 
```
def nothing %{
  echo -debug nothing
  %sh{sleep 1; date}
}
nothing
nothing
```

Why would the parsing of `nothing` get invalidated? Again, this not a deep parse.. OK, with `try`:

```
def nothing %{
  try %{ echo -debug nothing }
  %sh{sleep 1; date}
}
nothing
nothing
```

It seems reasonable that whenever the editor evaluates a codeblock, whether literal, or coming from a def, or from within a sub-try of a def, or whatever, it can lookup the parsetree.

**Comment by mralusw** [comment]
> I'd be surprised to see this change happen. Seems like a lot of complexity to fix a use case that probably goes against mawww's idea of what kakscript should be. 

Yes. Except you have `kakscript` re-parses all over the place. Every defined command gets re-parsed whenever the user types it in.

**Comment by mralusw** [comment]
>

**Comment by mralusw** [comment]
> What is a "parse tree" in this instance? It can't be a LISP-like AST, because Kakoune relies so heavily on expansions that any non-trivial AST would be invalidated near-instantly.

@Screwtapello I don't want to look yet into how kak's parsetree's are implemented, but I assume there must be some step between strings and execution.

Let me explain a bit more on invalidation:

```
def proc1 %{
  nop
  %sh{date}
  eval %{ echo -debug 1}
  try %{ %sh{date} }
}
```

- the parse tree for proc1 can be cached; it should include calls to (first) nop; then sh, eval and try, which are dynamically-evaluted. It won't be invalidated as long as proc1 isn't redefined.. It's a simple string to tree mapping.
- the eval block (`echo -debug 1`) is a single command node, that can be cached, and will not be invalidated
- the try block is a simple parsetree that consists of a single sh call, and will not be invalidated
- all the inner sh blocks produces a different (invalid) codeblock each time, which can be cached and LRU invalidated

In other words, the LRU cache is a mapping between strings and tokenized-and-parsed trees. It shouldn't matter what the "inner" trees are.

So I don't see any conceptual problems regarding hash-mapping strings to parsetrees. How the LRU cache would work / evict / have generations is an optimization story.

**Comment by eraserhd** [comment]
Hashing the string takes O(N), and parsing it also takes O(N).

**Comment by mralusw** [comment]
> Hashing the string takes O(N), and parsing it also takes O(N).

You hash once in O(N) so that you can use the hashed value (e.g. a hook) any number of times in O(1). But I suppose this is basic.

**Comment by mralusw** [comment]
@eraserhd sorry; I now see what you mean: strings have to be traversed, whether by the parser or the cache mechanism. Thus, hashing would only confer a constant factor speedup for each block (then again, the multiplier matters &mdash; if hashing is twice as fast as parsing, than that applies to all hooks).

Also, depending on the internal representation, wouldn't the string contents of a `def` be the same pointer (until redefined)? If so, the string doesn't need to be re-traversed. Same with constant-string `try`'s and `eval`'s (but not with dynamic strings: `eval %sh{...}`, `try "%opt{...}` etc)

**Comment by mawww** [comment]
Parsing in Kakoune does not really have anything that looks like a "parse tree" because Kakoune's grammar is actually very simple. From Kakoune's perspective there is only a list of tokens and those get executed as commands whenever we hit a command separator token.

The only thing we can cache (and I think this is what you are proposing) is the tokenization, I am not sure we would manage to gain much, except for highly commented code. Kakoune's tokenization logic is pretty simple, so I would not expect hashing to run circles around it, however it means we now need to hash first to lookup the cache, so we are impacting the performance of the first time we execute a specific string to maybe improve the performance of the following executions of that string.

If we were to hash based of the string "identity" instead of content, then that would essentially amount to replacing the string to execute with its pre-tokenized version. It would maybe improve performance at the expense of memory usage.

My performance-intuition is that we would not gain much in general use cases, but intuitions can be wrong, a POC pull request with performance mesurement would go a long way in proving/infirming the case for caching.
